File: Blind 75/07 - Trees/105-construct-binary-tree-from-preorder-and-inorder-traversal.py
'''
Given two integer arrays preorder and inorder where preorder is the preorder traversal of a binary tree and inorder is the inorder traversal of the same tree, construct and return the binary tree.
'''

from copy import deepcopy
from time import time
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(
        self, preorder: List[int], inorder: List[int]
    ) -> Optional[TreeNode]:
        def dfs(pre, inord):
            if not pre or not inord:
                return None

            node = TreeNode(pre[0])
            # Get inorder node position based on preorder node
            i = inord.index(pre[0])
            # Left of node uses inorder up to node position
            node.left = dfs(pre[1 : i + 1], inord[:i])
            # Right of node uses inorder starting from node position + 1
            node.right = dfs(pre[i + 1 :], inord[i + 1 :])

            return node

        return dfs(preorder, inorder)

    def reference(
        self, preorder: List[int], inorder: List[int]
    ) -> Optional[TreeNode]:
        if not preorder or not inorder:
            return None

        root = TreeNode(preorder[0])
        mid = inorder.index(preorder[0])
        root.left = self.reference(preorder[1 : mid + 1], inorder[:mid])
        root.right = self.reference(preorder[mid + 1 :], inorder[mid + 1 :])

        return root

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                # Create deep copy
                copy = deepcopy(case)
                if i == 0:
                    print(self.buildTree(*copy).__dict__)
                else:
                    self.buildTree(*copy)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                # Create deep copy
                copy = deepcopy(case)
                if i == 0:
                    print(self.reference(*copy).__dict__)
                else:
                    self.reference(*copy)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [([3, 9, 20, 15, 7], [9, 3, 15, 20, 7]), ([-1], [-1])]
    test.quantify(test_cases)

File: Blind 75/07 - Trees/JavaScript/105-construct-binary-tree-from-preorder-and-inorder-traversal.js
const { performance } = require('perf_hooks');

class TreeNode {
  constructor(val, left, right) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

class Solution {
  method(preorder, inorder) {
    if (preorder.length === 0 || inorder.length === 0) return null;

    const i = inorder.indexOf(preorder.shift());
    let root = new TreeNode(inorder[i]);
    root.left = this.method(preorder, inorder.slice(0, i));
    root.right = this.method(preorder, inorder.slice(i + 1));

    return root;
  }

  reference(preorder, inorder) {
    const isBaseCase = !preorder.length || !inorder.length;
    if (isBaseCase) return null;

    const { leftInorder, mid, rightInorder } = this.getPointers(
      preorder,
      inorder
    );
    const root = new TreeNode(inorder[mid]);

    root.left = this.reference(preorder, leftInorder);
    root.right = this.reference(preorder, rightInorder);

    return root;
  }

  getPointers = (preorder, inorder) => {
    const next = preorder.shift();
    const mid = inorder.indexOf(next);
    const leftInorder = inorder.slice(0, mid);
    const rightInorder = inorder.slice(mid + 1);

    return { leftInorder, mid, rightInorder };
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        // Create deep copy
        const copy = JSON.parse(JSON.stringify(input));
        if (i === 0) console.log(this.method(copy[0], copy[1]));
        else this.method(copy[0], copy[1]);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((run, i) => {
      testCases.map((input) => {
        // Create deep copy
        const copy = JSON.parse(JSON.stringify(input));
        if (i === 0) console.log(this.reference(copy[0], copy[1]));
        else this.reference(copy[0], copy[1]);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [
    [3, 9, 20, 15, 7],
    [9, 3, 15, 20, 7],
  ],
  [[-1], [-1]],
];
test.quantify(testCases);
